src/exporters/prometheus.rs

```diff
@@ -103,43 +112,90 @@ struct PowerMetrics {
     topology: Mutex<Topology>,
     last_request: Mutex<Duration>,
     qemu: bool,
+    containers: bool,
     hostname: String,
 }
 
-#[actix_web::main]
-/// Main function running the HTTP server.
+#[tokio::main]
 async fn runner(
     topology: Topology,
     address: String,
     port: String,
     suffix: String,
     qemu: bool,
+    containers: bool,
     hostname: String,
-) -> std::io::Result<()> {
-    if let Err(error) = address.parse::<IpAddr>() {
-        panic!("{} is not a valid ip address: {}", address, error);
-    }
-    if let Err(error) = port.parse::<u64>() {
-        panic!("Not a valid TCP port numer: {}", error);
-    }
-
-    HttpServer::new(move || {
-        App::new()
-            .data(PowerMetrics {
-                topology: Mutex::new(topology.clone()),
+) {
+    if let Ok(addr) = address.parse::<IpAddr>() {
+        if let Ok(port) = port.parse::<u16>() {
+            let socket_addr = SocketAddr::new(addr, port);
+            let power_metrics = PowerMetrics {
+                topology: Mutex::new(topology),
                 last_request: Mutex::new(Duration::new(0, 0)),
                 qemu,
+                containers,
                 hostname: hostname.clone(),
-            })
-            .service(web::resource(&suffix).route(web::get().to(show_metrics)))
-            .default_service(web::route().to(landing_page))
-    })
-    .workers(1)
-    .bind(format!("{}:{}", address, port))?
-    .run()
-    .await
+            };
+            let context = Arc::new(power_metrics);
+            let make_svc = make_service_fn(move |_| {
+                let ctx = context.clone();
```

Comment:
Here I would use shadow to keep the same name `let context = context.clone()`. I think it is easier to follow.
Idem for sfx below.
And do you really need to clone twice ? here and in the async block below ? I'm not sure but maybe it is only required in the async block.

Reply:
I had errors I couldn't resolve if I didn't clone twice. Maybe I did it wrong. I'll try to give you some details (next week most probably) so we can look into it and see if it should be done differently.